rfd/00TBD-webui-paginate-filter.md

```diff
@@ -0,0 +1,344 @@
+---
+authors: Lisa Kim (lisa@goteleport.com)
+state: draft
+---
+
+# RFD TBD - WebUI server-side paginating and filtering
+
+## What
+
+Provide server-side pagination and filtering capabilities for the web UI for select resources: nodes, apps, dbs, kubes, and desktops.
+
+## Why
+
+Currently, the web api calls upon non paginated endpoints that retrieves an entire list of resources which for higher ranges ~20k+, results in a ~30+ second load time for the UI. With pagination support, we can control the limit of resources to fetch, increasing load speed, and user experience. Filtering will then need to be done on the server-side to apply it to the entire list of resources.
+
+## Details
+
+### Pagination
+
+There is already pagination support for resources: nodes, apps, dbs. Kubes is [pending](https://github.com/gravitational/teleport/pull/9096#issuecomment-989291135)
+
+### Filter
+
+There will be filters that will be applied in order of precedence: RBAC > label > search > sort. RBAC will not be discussed since this check existed before. 
+
+### Filter: Label Query Language
+
+The query language for labels will be modeled after [kubernetes labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/). Kubernetes uses `set based` filtering which matches a label `key` against a set of `values` as shown in examples below.
+
+The language will be the same across the web ui and cli tools.
+
+Supported operators:
+- Comma between labels acts as an `AND` operator (same interpretation as current `tsh ls <label1>,<label2>`)
+- Pipe between label `values` denotes a set and acts as a `OR` operator
+- Minus before a label acts as a `NOT` operator
+
+Additionally, an empty right-hand side of the `equals` char means `any values`.
+
+| Format                        | Operator  | Example          | Description                                 |
+|-------------------------------|-----------|------------------|---------------------------------------------|
+| `key=value`                   | in        | `env=prod`       | rows with label `env=prod`                  |
+| `key=value1\|value2\|value*`  | in        | `env=prod\|dev`  | rows with labels `env=prod` or `env=dev`    |
+| `-key=value`                  | notin     | `-env=prod`      | rows without label `env=prod`               |
+| `-key=value1\|value2\|value*` | notin     | `-env=prod\|dev` | rows without labels `env=prod` or `env=dev` |
```

Comment:
The preceeding `-` might confuse our CLI parsing library, since it allows using `-option` and `--option` interchangeably.  Kubernetes uses `key!=value` for "notin" instead, maybe we could go with something like that?  If we did use `!=`, I'd suggest taking another page from the kubernetes playbook and allowing `=` and `==` to be used interchangeably, since a lot of folks have an established habit of `==`/`!=` being used for equality/inequality.